File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
"""
Time = O(n)
Space = O(n)
"""
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if token in "+-*/":
                opd2 = int(stack.pop())
                opd1 = int(stack.pop())
                if token == "+":
                    stack.append(opd1 + opd2)
                elif token =="-":
                    stack.append(opd1 - opd2)
                elif token == "*":
                    stack.append(opd1 * opd2)
                else:
                    stack.append(int(opd1 / opd2))
            else:
                stack.append(int(token))
        
        return stack[0]
            

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.cpp
#include <stack>

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;

        for (string token: tokens) {
            if (token.size() > 1 || isdigit(token[0])) {
                st.push(stoi(token));
            } else {
                int op2 = st.top();
                st.pop();
                int op1 = st.top();
                st.pop();

                switch (token[0]) {
                    case '+':
                        st.push(op1 + op2);
                        break;
                    case '-':
                        st.push(op1 - op2);
                        break;
                    case '*':
                        st.push(op1 * op2);
                        break;
                    case '/':
                        st.push((int) (op1 / op2));
                        break;
                }
            }
        }

        return st.top();
    }
};